### DISCUSSION TITLE
RLS based on relationships?

### DISCUSSION THREAD

**Discussion by tconroy** [supabase/supabase#2337] [Questions]
I'm new to RLS with Postgres and having some trouble accomplishing what I'm trying to do.

Basically: I have an app where there are Users that can have a membership to an Organization. The tables look something like this:

```
-- users
+---------+--------------+
| id (PK) |  full_name   |
+---------+--------------+
| uuid()  | varchar      |
+---------+--------------+

-- organizations
+---------+---------+
| id (PK) |  name   |
+---------+---------+
| uuid()  | varchar |
+---------+---------+

-- organization_memberships
+---------+----------------------+--------------+
| id (PK) | organization_id (FK) | user_id (FK) |
+---------+----------------------+--------------+
| uuid()  | uuid()               | uuid()       |
+---------+----------------------+--------------+
```

So, the `organization_memberships` table maps `user_id` to `organizations`.

I want to configure RLS on the `organizations` table, so that a user _can only be returned organizations which they belong to_. 

This is my current attempt, but it doesn't seem to work. I'm getting back `null` for my `organizations` query:

```
(
  EXISTS (
    SELECT
      (
        organizations.id IN (
          SELECT
            om.organization_id
          FROM
            (
              organization_memberships om
              JOIN users u ON ((om.user_id = uid()))
            )
        )
      )
  )
)
```

**Comment by tconroy** [comment]
Just an update.. 

I tried a slightly different format, but still not working:
```
ALTER POLICY "User can only Select organizations where they are member" ON public.organizations FOR SELECT USING (
  auth.uid() IN (
    SELECT user_id
    FROM organization_memberships
    WHERE organization_memberships.organization_id = organization.id
  )
);
```

Similar to the previous attempt, this causes no `organization` objects to be returned, despite the user existing in the `organization_memberships` table. 


Not really sure what I'm doing wrong here... I inspected the JWT and confirmed that the desired user ID is present under the `payload.sub` field.

**Comment by fishnux** [comment]
Did you end up solving this issue? I'm not a Supabase user yet, just curious as this would be an important feature on my app. Curiously, Supabase's website has Organization support, so I wonder how they do it? (if they use Supabase on the back-end)

**Reply by GaryAustin1** [reply]
I like using a function for these kinds of cases to avoid worrying about the RLS policy being enforced on a table in the policy query which will cause problems.  Plus I can use the same function in other table policies.

https://supabase.com/docs/guides/auth/row-level-security#policies-with-security-definer-functions

**Reply by steve-chavez** [reply]
> https://supabase.com/docs/guides/auth/row-level-security#policies-with-security-definer-functions

Hm, we need to improve those docs to create the security definer function in a "private" schema, being in "public" could leak data.

**Comment by joshenlim** [comment]
Hey @tconroy, I've replied to you in the Github issue [here](https://github.com/supabase/supabase/issues/180) but also pasting my answer here for visibility

> I've tried this policy, but I'm getting an error, Error updating policy: missing FROM-clause entry for table "organization":

There's a typo in your policy's USING statement - your table's name is `organizations` but you're doing the check on `organization.id`

> I'm still seeing the same issue, though: Getting 0 results when querying for organiziations that a user belongs to via a organzation_memberships table.

Hmm i'm actually unable to reproduce this unfortunately, here's how I reproduced your environment, maybe you could check if you might be missing any steps

1. Created the tables you mentioned and added some fake data
<img width="967" alt="image" src="https://user-images.githubusercontent.com/19742402/166669496-0c8be9a1-80b7-4164-b3c9-8d5df2ebcfb3.png">

2. Set up the policy on the `organizations` table and enabled RLS on the table
<img width="1182" alt="image" src="https://user-images.githubusercontent.com/19742402/166669603-32375913-9a77-45e1-8ba3-28683c92e181.png">

3. Set up a demo app on codesandbox ([ref](https://codesandbox.io/s/stupefied-khayyam-8myqyd?file=/src/App.js))
I've added the code sandbox as a reference - it may/may not run successfully as I created a demo project just to test it, which I'm planning to remove soon
![image](https://user-images.githubusercontent.com/19742402/166669816-d021d64a-876d-4ac1-8d26-2f8027afa070.png)

4. Create a user via the demo app and also save the user in the `users` table of the `public` schema (under the `full_name` "Amanda")
<img width="1130" alt="image" src="https://user-images.githubusercontent.com/19742402/166669908-0c0c862b-a4f7-4142-893d-6e1d7d414cea.png">
<img width="947" alt="image" src="https://user-images.githubusercontent.com/19742402/166670006-5aab2e8a-0f21-4b08-8112-efeda9039d16.png">

5. Manually added a record in `organization_memberships` to link "Amanda" to "Company A"
<img width="1009" alt="image" src="https://user-images.githubusercontent.com/19742402/166670063-1e1e3607-f128-4a43-9ac1-7f2258004458.png">

6. Check on demo app to do a `select` on the `organizations` table while logged in as "Amanda", in which I correctly only receive 1 result instead of 4
![image](https://user-images.githubusercontent.com/19742402/166670150-1157fb1e-9f63-4f3e-8a53-785135876880.png)

This is the code I ran:
```
const retrieveOrganizations = async () => {
    const { data } = await supabase.from("organizations").select("*");
    console.log("Organizations", data);
};
```

Hope this helps!
